feat(electrum)!: Update `bdk_electrum` to use merkle proofs

LagginTimes opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->
Fixes #980.

### Description

This PR is the first step in reworking `bdk_electrum` to use merkle proofs. When we fetch a transaction, we now also obtain the merkle proof and block header for verification. We then insert an anchor only after validation that the transaction exists in a confirmed block. The loop logic that previously existed in `full_scan` to account for re-orgs has also been removed as part of this rework.

This is a breaking change because `graph_update`s now provide the full `ConfirmationTimeHeightAnchor` type. This removes the need for the `ElectrumFullScanResult` and `ElectrumSyncResult` structs that existed only to provide the option for converting the anchor type from `ConfirmationHeightAnchor` into `ConfirmationTimeHeightAnchor`.

### Notes to the reviewers

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->
* `ConfirmationTimeHeightAnchor` and `ConfirmationHeightAnchor` have been removed.
* `ConfirmationBlockTime` has been introduced as a new anchor type.
* `bdk_electrum`'s `full_scan` and `sync` now return `graph_update`s with `ConfirmationBlockTime`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature

LagginTimes committed:
feat(electrum)!: Update `bdk_electrum` to use merkle proofs

notmandatory commented:
Since this doesn't touch the wallet APIs I moved it to the beta milestone.

LagginTimes commented:
> Since this doesn't touch the wallet APIs I moved it to the beta milestone.

I'm okay with this!

LLFourn commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -371,120 +287,90 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+            .inner
+            .transaction_get_merkle(&txid, confirmation_height as usize)
+        {
+            let header = self.inner.block_header(merkle_res.block_height)?;
```
I think we should have a `block_header_cache` here too. Then you can get the last block header we had at that height. If the merkle proof validation fails we can fetch it again, and validate against the new one. This will reduce the number of requests we make to electrum by a bit.

LLFourn commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -371,120 +287,90 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+            if is_confirmed_tx {
+                let _ = graph_update.insert_anchor(
+                    txid,
+                    ConfirmationTimeHeightAnchor {
```
This seems like this is a suboptimal anchor to use since you're entering the same `block_height` twice. It would actually be good if you could delete `ConfirmationTimeHeightAnchor` and `ConfirmationHeightAnchor` from the codebase now since both electrum and esplora will report the exact block that the tx is in with this PR. It only needs to be recorded once.

For now this would be good to use for both:

```rust
struct ConfirmationBlockTime {
    block_id: BlockId,
    confirmation_time: u64
}
```

Later we can add the block position to this and maybe even the proof. 

Deleting `ConfirmationTimeHeight` anchor would need ACK from @evanlinjin  and @notmandatory since it's used by `Wallet`. Just for background the only reason `ConfirmationTimeHeightAnchor` exists is because of how wacky our electrum client was before this PR. I think simplifying the anchors to make use of this should be done ASAP.

LLFourn replied:
If we don't want to delete the old anchors we should still return the correct thing from here. It can be `Into` converted into the old one later.

evanlinjin replied:
ACK

notmandatory replied:
I support simplifying the anchors now. I'm happy to do whatever changes are required for to the sqlite storage crate. 

cc: @matthiasdebernardini this anchor change will also impact the postgres store you're working on.

LLFourn commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -371,120 +287,90 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+            .transaction_get_merkle(&txid, confirmation_height as usize)
+        {
+            let header = self.inner.block_header(merkle_res.block_height)?;
+            let is_confirmed_tx = electrum_client::utils::validate_merkle_proof(
```
Is it possible to use rust bitcoin's merkle proof validator for this? Why does electrum have its own validation logic?

LagginTimes replied:
The validation logic being used right now was implemented as a utility in `rust-electrum-client`, which was needed for LDK's electrum integration:
https://github.com/bitcoindevkit/rust-electrum-client/pull/122

Would using rust bitcoin be more advantageous in this scenario?

LLFourn replied:
I have no idea. I am just curious why @tnull added that when there is presumably this algorithm is implemented in `rust-bitcoin`.

tnull replied:
IIRC, Electrum's `get_merkle`/`GetMerkleRes` doesn't provide sufficient/slightly different data, so that we couldn't construct `bitcoin::MerkleBlock`/ `PartialMerkleTree` to use upstream's validation logic (which of course would have been our preferred way).

LLFourn reviewed (commented):
Thanks this simplifies a lot. Left a few comments around the anchors/merkle proof bit.

notmandatory commented:
@LagginTimes oops, I didn't read all the comments before, looks like the wallet API will need to be updated so I've moved this back to the alpha milestone.

LagginTimes committed:
feat(chain)!: Implement `ConfirmationBlockTime`

Both `bdk_electrum` and `bdk_esplora` now report the exact block
that the transaction is in, which removes the need for having the
old `ConfirmationTimeHeightAnchor` and `ConfirmationHeightAnchor`.
This PR introduces a new, simpler anchor type that can be modified
to support additional data in the future.

ValuedMammal commented on crates/chain/src/chain_data.rs:
```diff
@@ -156,70 +156,27 @@ impl From<(&u32, &BlockHash)> for BlockId {
-    ///
-    /// It is assumed that this value is never larger than the height of the anchor block.
-    pub confirmation_height: u32,
+pub struct ConfirmationBlockTime {
```
docs: this line no longer applies

> Note that the confirmation block and the anchor block can be different here.

ValuedMammal commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -944,9 +944,12 @@ fn test_chain_spends() {
+                block_id: BlockId {
+                    hash: tip.get(ht).unwrap().hash(),
+                    height: ht,
+                },
```
```suggestion
                block_id: tip.get(ht).unwrap().block_id(),
```

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -88,87 +120,32 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+        let mut graph_update = TxGraph::<ConfirmationBlockTime>::default();
+        let mut last_active_indices = BTreeMap::<K, u32>::new();
+
+        for (keychain, keychain_spks) in request.spks_by_keychain {
```
```suggestion
        for (keychain, spks) in request.spks_by_keychain {
```

ValuedMammal replied:
📌  The docs for `full_scan` are a bit lacking with respect to `fetch_prev_txouts`, see https://github.com/bitcoindevkit/bdk/pull/1494#issuecomment-2204199786

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -227,70 +203,46 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
-        graph_update: &mut TxGraph<ConfirmationHeightAnchor>,
-        spks: &mut impl Iterator<Item = (I, ScriptBuf)>,
+        graph_update: &mut TxGraph<ConfirmationBlockTime>,
+        mut spks: impl Iterator<Item = (I, ScriptBuf)>,
```
docs: "cps" and "tx_cache" are no longer relevant parameters, same for `populate_with_outpoints` below

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -371,120 +317,100 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+                );
+            }
+        }
+        Ok(())
```
question: Is there a scenario where the `validate_merkle_for_anchor` function could fail and we'd want to know about it? in other words should we be worried if validating the proof fails even after updating the header?

notmandatory replied:
📌  This is a good question and it makes sense to have a runtime option or feature to throw an error if merkle proof validation fails. But I don't think this has to be done in this PR or for the beta release since it will only change the API for the `bdk_electrum` crate. @ValuedMammal would you mind just making an issue for this?

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -527,6 +453,7 @@ fn construct_update_tip(
 
     let new_tip = new_blocks
+        .clone()
         .into_iter()
```
if we just use `.iter()` we can remove the call to `clone`

ValuedMammal commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -155,13 +152,11 @@ fn tx_can_become_unconfirmed_after_reorg() -> anyhow::Result<()> {
-            false,
-        )?
-        .with_confirmation_time_height_anchor(&client)?;
+    let update = client.sync(
+        SyncRequest::from_chain_tip(recv_chain.tip()).chain_spks([spk_to_track.clone()]),
+        5,
+        false,
+    )?;
```
We should be able to make assertions on the expected graph anchors, something like:

```diff
     // Create transactions that are tracked by our receiver.
+    let mut txids = vec![];
+    let mut blockhashes = vec![];
     for _ in 0..REORG_COUNT {
-        env.send(&addr_to_track, SEND_AMOUNT)?;
-        env.mine_blocks(1, None)?;
+        txids.push(env.send(&addr_to_track, SEND_AMOUNT)?);
+        blockhashes.extend(env.mine_blocks(1, None)?);
     }
 
     // Sync up to tip.
     env.wait_until_electrum_sees_block()?;
     let update = client.sync(
@@ -163,10 +161,17 @@ fn tx_can_become_unconfirmed_after_reorg() -> anyhow::Result<()> {
         .map_err(|err| anyhow::anyhow!("LocalChain update error: {:?}", err))?;
     let _ = recv_graph.apply_update(update.graph_update.clone());
 
     // Retain a snapshot of all anchors before reorg process.
     let initial_anchors = update.graph_update.all_anchors();
+    let anchors: Vec<_> = initial_anchors.iter().cloned().collect();
+    assert_eq!(anchors.len(), REORG_COUNT);
+    for i in 0..REORG_COUNT {
+        let (anchor, txid) = anchors[i];
+        assert_eq!(anchor.anchor_block().hash, blockhashes[i]);
+        assert_eq!(txid, txids[i]);
+    }
```

ValuedMammal reviewed (commented):
I did a first pass, mostly documentation nits. Sorry I was reviewing while you were pushing, so comments might be outdated

LagginTimes committed:
test(electrum): Imported `bdk_esplora` tests into `bdk_electrum`

notmandatory commented:
@LagginTimes hey sorry I just merged some smaller PRs and now this one needs a rebase.

evanlinjin commented:
I think we should not wait for complete tests before merging this PR. As long as the examples work fine it is good. This is for "beta" which does not need to be bug free. Also having this PR merged means we can work on the other anchor changes (which I need to create a ticket for).

evanlinjin commented:
Also, the anchor changes that I am suggesting is going to conflict with `ConfirmationBlockTime`.

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -190,117 +167,87 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
-                    if unused_spk_count > stop_gap {
-                        return Ok(scanned_spks);
+                    unused_spk_count = unused_spk_count.saturating_add(1);
+                    if unused_spk_count >= stop_gap {
```
💯

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -527,10 +449,13 @@ fn construct_update_tip(
         // Prune `new_blocks` to only include blocks that are actually new.
-        .filter(|(height, _)| Some(*height) > agreement_height)
-        .map(|(height, hash)| BlockId { height, hash })
+        .filter(|(height, _)| Some(*<&u32>::clone(height)) > agreement_height)
```
nit, feel free to ignore

```suggestion
        .filter(|(&height, _)| Some(height) > agreement_height)
```

ValuedMammal commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -19,6 +21,222 @@ fn get_balance(
 }
 
+#[test]
+pub fn test_update_tx_graph_without_keychain() -> anyhow::Result<()> {
```
📌 If I recall, the name of this test is a relic from before the `sync` and `full_scan` concepts were developed. I might now call it `sync_should_update_chain_and_tx_graph` or similar, but not terribly important at the moment.

ValuedMammal reviewed (approved):
ACK 1a62488abfd6213c2780c1465aebe01dfbb6b20a

notmandatory reviewed (approved):
ACK 1a62488abfd6213c2780c1465aebe01dfbb6b20a

Thanks for also adding the extra esplora tests for electrum!
